add error checks on intrinsic tfuncs

aviatesk opened the pull request:
This is useful for `AbstractInterpreter`-based static analyzer like JET.jl.
May come with some compilation performance cost, so I will keep an eye on it.

@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/48000) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/110ca93_vs_5bd0160/report.md).

aviatesk committed:
allow external `AbstractLattice` to customize intrinsic tfuncs

This would be useful for `AbstractInterpreter`-based static analyzers
that use custom `AbstractLattice`.

aviatesk commented:
Okay, so there seems to be some compilation performance cost. I don't think it's very useful to turn on this error check for native compilation, so I will just add changes that allow external `AbstractLattice` to customize intrinsic tfuncs instead of implementing the error checks for `JLTypeLattice`.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/48000#issuecomment-1365590052) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/37d04c2_vs_5c493c5/report.md).
